cache/reference_counted_test.go

```diff
@@ -13,88 +13,100 @@ import (
 )
 
 func TestCacheWithDisposal(t *testing.T) {
-	const (
-		size        = 7
-		parallelism = 100
-		repeats     = 2000
-	)
 	type record struct {
 		key      int
 		disposed int32
 		live     int32
 	}
 	ctx := context.Background()
-	var elements [size]*record
-	ch := make(chan error, parallelism*2)
-	allErrs := make(chan []error)
-	go func() {
-		errs := make([]error, 0)
-		for err := range ch {
-			errs = append(errs, err)
-		}
-		allErrs <- errs
-		close(allErrs)
-	}()
-	p := cache.ParamsWithDisposal{
-		Name:   t.Name(),
-		Logger: logging.Default().WithField("testing", true),
-		Shards: 3,
-		Size:   size,
-		OnDispose: func(v interface{}) error {
-			element := v.(*record)
-			if n := atomic.AddInt32(&element.disposed, 1); n != 1 {
-				ch <- fmt.Errorf("%d disposals of %+v", n, *element)
-			}
-			if l := atomic.CompareAndSwapInt32(&element.live, 1, 0); !l {
-				ch <- fmt.Errorf("disposal of already-dead %+v", *element)
-			}
-			return nil
-		},
+
+	cases := []struct {
+		name                       string
+		size, parallelism, repeats int
+		delay                      time.Duration
+	}{
+		{name: "full speed", size: 7, parallelism: 100, repeats: 2000},
+		{name: "try to race", size: 7, parallelism: 9, repeats: 500, delay: time.Millisecond},
 	}
-	c := cache.NewCacheWithDisposal(p)
 
-	numCreated := int32(0)
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			var elements = make([]*record, tc.size)
+			ch := make(chan error, tc.parallelism*2)
+			allErrs := make(chan []error, tc.parallelism*2)
+			go func() {
+				errs := make([]error, 0)
+				for err := range ch {
+					errs = append(errs, err)
+				}
+				allErrs <- errs
```

Comment:
Why do you need to error channels?

Reply:
Leftover from when we thought goroutines were not allowed to issue errors.  SInce they're only prohibited from failing, removed!  Thanks!